src/core/Akka.Remote/Transport/Streaming/StreamAssociationHandle.cs

```diff
@@ -0,0 +1,238 @@
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using Akka.Actor;
+using Google.ProtocolBuffers;
+
+namespace Akka.Remote.Transport.Streaming
+{
+    public class StreamAssociationHandle : AssociationHandle
+    {
+        private readonly Stream _stream;
+        private readonly AsyncQueue<ByteString> _writeQueue;
+        private IHandleEventListener _eventListener;
+
+        private readonly byte[] _readBuffer;
+        private readonly byte[] _writeBuffer;
+        private readonly TimeSpan _flushWaitOnShutdown;
+
+        private readonly TaskCompletionSource<bool> _initialized;
+        private readonly TaskCompletionSource<bool> _stopped;
+        private readonly Task _writeLoop;
+
+        /// <summary>
+        /// Returns true if gracefully stopped, otherwise false.
+        /// </summary>
+        public Task<bool> Stopped => _stopped.Task;
+
+        public StreamAssociationHandle(StreamTransportSettings settings, Stream stream, Address localAddress, Address remoteAddress)
+            : base(localAddress, remoteAddress)
+        {
+            _stream = stream;
+            _writeQueue = new AsyncQueue<ByteString>();
+
+            _readBuffer = new byte[settings.StreamReadBufferSize];
+            _writeBuffer = new byte[settings.StreamWriteBufferSize];
+            _flushWaitOnShutdown = settings.FlushWaitTimeout;
+
+            _initialized = new TaskCompletionSource<bool>();
+            _stopped = new TaskCompletionSource<bool>();
+            _writeLoop = Task.Run(() => WriteLoop());
+        }
+
+        internal void Initialize(IHandleEventListener eventListener)
```

Comment:
These are async methods, so you can view them as logical thread, they do not block a real thread. WriteAsync/ReadAsync use the IO threadpool.

The thread is released when await is called on an uncompleted task. When the task completes, the loop continue on the thread that completed the task. Might be an IO completion thread if coming back from ReadAsync/WriteAsync or might be a threadpool thread if coming back from the AsyncQueue.